### ISSUE TITLE
Make createElement(undefined) warning more descriptive

### ISSUE THREAD

**Issue by gaearon** [facebook/react#7307]
**Update: claimed by @jin**

This is what I see when I mistype an import:

<img width="768" alt="screen shot 2016-07-19 at 21 12 01" src="https://cloud.githubusercontent.com/assets/810438/16964870/7392ba8c-4df5-11e6-8af6-7d00181e5089.png">

This is not very useful.

When `type` is `undefined`, we should provide a better message. In 95% of cases it is caused by an invalid import. We should create a page explaining common cases how this could happen (e.g. mismatching default/named export, forgetting to export the component, or importing a non-existing named export), and link to that page from the warning.

**Comment by gaearon** [comment]
If you intend to work on this, please write in a comment here. The change to the code is very small so most of the effort would be spent on creating a good warning page explaining common problems and solutions.

Here is an example of such warning page: https://github.com/facebook/react/blob/master/docs/warnings/legacy-factories.md. This one would be similar (but about a different topic).

**Comment by jin** [comment]
Can I look into this?

**Comment by gaearon** [comment]
Sure!

**Comment by gaearon** [comment]
I’m “assigning” myself so I keep track of this issue but it’s yours.

**Comment by Pajn** [comment]
Please include a stacktrace!
This is the worst possible message to get after doing a big merge as there are absolutely nothing to guide you to where the problem can be.

**Comment by jin** [comment]
@Pajn Will do! I'm bitten by the lack of information many times too, especially when dealing with HOC.

**Comment by zpao** [comment]
Similar: #7215.

**Comment by gaearon** [comment]
@jin Any updates? If you’re too busy, it’s fine—just let us know and we’ll ask somebody else to look into this!

**Comment by jin** [comment]
@gaearon Really sorry for losing track of this - please free this up for someone else.

**Comment by gaearon** [comment]
@jin Anything we can help you with? This should only involve [changing these few lines](https://github.com/facebook/react/blob/c3ce0f24c98917ead456ab56b00360ed140ad371/src/isomorphic/classic/element/ReactElementValidator.js#L189-L197). Also would love to hear your feedback on the confusing parts of the codebase.

**Comment by alexzherdev** [comment]
I can pick this up if you guys decide this needs a fresh pair of eyes.

**Comment by jin** [comment]
Actually, I'm done with changing that warning to link to a warning docs page. Right now, I'm primarily blocked with coming up with the different examples/scenarios that the issue can occur in, on top of the ones you've mentioned.

The codebase isn't confusing per se, it's the part of transforming my understand from actual applications of React to the underlying implementations of them.

**Comment by jin** [comment]
@gaearon done! Happy to hear any feedback.

**Comment by Wilfred** [comment]
As discussed in #7485, I think we should also report /what/ value we saw, rather than just a list of forbidden types. Compare

> Warning: React.createElement: type should not be null, undefined, boolean, or number. It should be a string (for DOM elements) or a ReactClass (for composite components).

with:

> Warning: React.createElement: 5 is an invalid value for type. It should be a string (for DOM elements) or a ReactClass (for composite components).

**Comment by jin** [comment]
Hi all, really really sorry for the delayed response: I'm moving between cities between these couple of weeks, haven't had the chance to sit down to address the feedback yet.

**Comment by Slruh** [comment]
@jin, have you made any progress on this issue? If you are too busy, I would love to take a stab at it.

**Comment by jin** [comment]
@Slruh Yep! There's some progress here: https://github.com/facebook/react/pull/7402 

I'm in the midst of addressing @gaearon's comments, and feel free to suggest any improvements!

**Comment by Slruh** [comment]
I'll take a look

**Comment by gravitypersists** [comment]
For me, this error is almost always a chore in finding the component that it's referring to. It'd help if the error message did better to identify that component. Maybe it'd help to display everything known about the element? Such as props, or the location of the element in the component hierarchy (perhaps a parent)?

> Warning: React.createElement: x is an invalid value for type. It should be a string (for DOM elements) or a ReactClass (for composite components).
```
<App>
    <Rooter>
        <div>
            <x>
              ^
```

Is something like that feasible and desirable?

**Comment by syranide** [comment]
@gravitypersists It just needs to explicitly say that it is undefined (rather than some obscure error) when `React.createElement` is called.

**Comment by joelseq** [comment]
Is this still being worked on or can I maybe try to take a stab at it?

**Comment by jin** [comment]
@joelseq I totally dropped the ball on this one, sorry! Please take over if
you'd like to!

**Comment by churchie317** [comment]
Looks like there hasn't been any activity for a few weeks. Anyone mind if I claim this for now?

**Comment by gaearon** [comment]
I think this was fixed in #8612, and being further improved in #8495.
Thanks for the offer to help though!
